src/python/doc/conf.py

```diff
@@ -8,18 +8,21 @@
 sys.path.insert(0, os.path.abspath("."))
 
 extensions = [
+    "sphinx.ext.napoleon",
+    # let napoleon be the first extension, cf. https://github.com/tox-dev/sphinx-autodoc-typehints/issues/15
     "matplotlib.sphinxext.plot_directive",
     "sphinx.ext.autodoc",
+    "sphinx_autodoc_typehints",
     "sphinx.ext.doctest",
     "sphinx.ext.mathjax",
     "sphinx.ext.ifconfig",
     "sphinx.ext.viewcode",
-    "sphinx.ext.napoleon",
     "sphinxcontrib.bibtex",
     "sphinx_paramlinks",
 ]
 
 autodoc_class_signature = "separated"
+autodoc_typehints = "none"
```

Comment:
Looking at https://github.com/tox-dev/sphinx-autodoc-typehints#options, it suggests we should add `napoleon_use_rtype = True`. It has 0 effect currently, but it might become relevant when we add more annotations?

`always_use_bars_union = True` looks a bit prettier than the default (False), but that's subjective.